datafusion/optimizer/src/type_coercion.rs

```diff
@@ -98,6 +101,81 @@ impl ExprRewriter for TypeCoercionRewriter<'_> {
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         match expr {
+            Expr::IsTrue(expr) => {
+                let expr = is_true(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsNotTrue(expr) => {
+                let expr = is_not_true(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsFalse(expr) => {
+                let expr = is_false(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsNotFalse(expr) => {
+                let expr =
+                    is_not_false(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::Like {
+                negated,
+                expr,
+                pattern,
+                escape_char,
+            } => {
+                let left_type = expr.get_type(&self.schema)?;
+                let right_type = pattern.get_type(&self.schema)?;
+                let coerced_type =
+                    coerce_types(&left_type, &Operator::Like, &right_type)?;
+                let expr = Box::new(expr.cast_to(&coerced_type, &self.schema)?);
+                let pattern = Box::new(pattern.cast_to(&coerced_type, &self.schema)?);
+                let expr = Expr::Like {
+                    negated,
+                    expr,
+                    pattern,
+                    escape_char,
+                };
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::ILike {
+                negated,
+                expr,
+                pattern,
+                escape_char,
+            } => {
+                let left_type = expr.get_type(&self.schema)?;
+                let right_type = pattern.get_type(&self.schema)?;
+                let coerced_type =
+                    coerce_types(&left_type, &Operator::Like, &right_type)?;
+                let expr = Box::new(expr.cast_to(&coerced_type, &self.schema)?);
+                let pattern = Box::new(pattern.cast_to(&coerced_type, &self.schema)?);
+                let expr = Expr::ILike {
+                    negated,
+                    expr,
+                    pattern,
+                    escape_char,
+                };
+                expr.rewrite(&mut self.const_evaluator)
+            }
+            Expr::IsUnknown(expr) => {
+                // will convert the binary(expr,IsNotDistinctFrom,lit(Boolean(None));
+                let left_type = expr.get_type(&self.schema)?;
+                let right_type = DataType::Boolean;
+                let coerced_type =
+                    coerce_types(&left_type, &Operator::IsNotDistinctFrom, &right_type)?;
```

Comment:
Is there a reason this isn't `Operator::IsUnknown`?

Reply:
I think you can got the reason from 
https://github.com/apache/arrow-datafusion/pull/3301
https://github.com/apache/arrow-datafusion/pull/3246
The unknown will be convert the physical binary op with `DistinctFrom` operation.
cc @alamb